### ISSUE TITLE
ALTER-DELETE query seemingly never completes

### ISSUE THREAD

**Issue by mdonkers** [ClickHouse/ClickHouse#7733]
**Describe the bug or unexpected behaviour**
In our production cluster we needed to remove certain records. To prevent overloading the cluster with lots of insertions happening, we used the following setup:
- Deletes were executed from 'master' nodes in the cluster, targeting the local (replicated) table
- Rows to be deleted were grouped using the `_part` column, and only a single 'part' should have been affected per statement
- The deletes were executed from a script, only continuing with the next 'part' when the mutation was finished.

We found at some point the mutations got stuck on every node. The mutation would not complete (`is_done = 0`) nor would it result in a failure. Records were removed though, and also the `system.part_log` table showed a new part was written, replacing the old 'part'.

ClickHouse logs would only show this (new mutation started when some was manually killed using `kill mutation where ...`):
```
clickhouse-server.log:2019.11.11 10:03:53.942679 [ 78 ] {} <Information> cust.calls_v2 (ReplicatedMergeTreeQueue): Loading 1 mutation entries: 0000000006 - 0000000006
clickhouse-server.log:2019.11.11 10:03:56.504414 [ 73 ] {} <Information> cust.calls_v2 (ReplicatedMergeTreeQueue): Loading 1 mutation entries: 0000000007 - 0000000007
clickhouse-server.log:2019.11.11 10:10:26.482707 [ 54 ] {} <Information> cust.calls_v2 (ReplicatedMergeTreeQueue): Loading 1 mutation entries: 0000000008 - 0000000008
```
Not sure if a coincidence or not, but it looked like the mutation always got stuck with `parts_remaining = 44` (while on average the DELETE had to process > 700 parts).

**How to reproduce**
We could only replicate this in our production cluster, other environments completed without issues but they don't have multi-node (replicated) setups.
* Which ClickHouse server version to use - ClickHouse server version 19.11.8 revision 54423
* Which interface to use, if matters - CLI client, using `--query` parameter
* `CREATE TABLE` statements for all tables involved - see below
* Queries to run that lead to unexpected result - see below

(redacted) query:
```
clickhouse client --database=cust --query="ALTER TABLE calls_v2 DELETE WHERE ${CALL_QUERY_WHERE} AND call_id IN (${CALLS})"
```
- where `${CALLS}` is grouping respective calls by a single `_part`
- The `${CALL_QUERY_WHERE}` contains filters on the primary keys for the table

(redacted) Create table:

```
CREATE TABLE cust.calls_v2 (
    app_id String,
    call_id String,
    t UInt64,
    retained_until_timestamp UInt64,
    cust_level UInt8,
...
)
ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/cust/calls_v2', '{replica}')
PARTITION BY toRelativeDayNum(toDate(retained_until_timestamp / 1000))
ORDER BY (app_id, cust_level, t)
SETTINGS index_granularity = 8192
```

**Expected behavior**
Mutations for the `ALTER .. DELETE` statements would finish.

**Error message and/or stacktrace**
None

**Comment by den-crane** [comment]
Check that all parts are not frozen (not RO) (https://github.com/ClickHouse/ClickHouse/issues/7152)

They should have write flags `drwxr-x---`
Sometimes CH leaves some parts with RO `dr--r-x---` without obvious reason (accidental bug).

**Comment by mdonkers** [comment]
Thanks,

I've checked all nodes, and while I do see differences with the column-files within the parts folders (some have `-r--r-----` but others have `-rw-r-----`), all parts folders seem to have the correct rights: all `drwxr-x---`.

Not sure if by killing and continuing this eventually resolved itself? But at least currently it doesn't show 'frozen' parts.

**Comment by den-crane** [comment]
Check result of this query

select * from system.parts where bytes_on_disk > (select toInt64(value) from system.merge_tree_settings where name = 'max_bytes_to_merge_at_max_space_in_pool')

It could be another reason why mutations stuck -- size of part  is bigger than max_bytes_to_merge_at_max_space_in_pool

**Comment by mdonkers** [comment]
Did not think about mentioning this. But yes, I had found that query on another issue thread and had executed it. Also checking now, we don't have parts bigger than the threshold that cannot get merged.

Perhaps not sufficiently emphasized, but as mentioned above, we found even when the mutation was 'stuck', the new parts resulting from deleting the records were written. The `system.part_log` showed the new part, with `merged_from` column containing the old part from which records were deleted. Also querying for these specific `call_id`s showed the records non-existing.
Not sure if this is all in line with the above known issues?
I'm thinking if there might have been an issue with folder permissions, would these then get cleaned up, or should I have found traces on this?

One last observation; the replica's followed suit. So as we ran cleanup only from the 'master' nodes, when we managed to get everything cleaned up there, replica's also showed 0 remaining records (using same filters).

Hope this provides any useful hints.

**Comment by sunadm** [comment]
I have the same problem, but I can not confirm the conditions is totally same.

Not sure, but check this issue may have some help:
#3953

**Comment by mdonkers** [comment]
Thx for the hint. Have been thinking about this, and have my doubts its the same issue.
The 'deletes' are not depending on another, and the script makes sure only one `DELETE` query runs per node. The only thing that might happen because we manually killed 'stuck' mutations, is that the script would trigger a `DELETE` for the same part. If there was something lingering in the system, it might get stuck to the same point?
But as we only started killing mutations once they got stuck, the question remains why that happened in the first place.
